Plugins/Public/base_plugin/PlayerCommands.cpp

```diff
@@ -451,6 +456,235 @@ namespace PlayerCommands
+			return;
+		}
+
+		wstring theaffiliation = HkGetWStringFromIDS(Reputation::get_name(tag)).c_str();
+		if (theaffiliation == L"Object Unknown")
+		{
+			PrintUserCmdText(client, L"ERR Undefined faction");
+			return;
+		}
+		base->ally_factions.insert(tag);
+
+		base->Save();
+		PrintUserCmdText(client, L"OK");
+	}
+
+	void BaseClearAllyFac(uint client, const wstring &args)
+	{
+		PlayerBase *base = GetPlayerBaseForClient(client);
+		if (CheckForBase(base, client)) return;
+		base->ally_factions.clear();
+		base->Save();
+		PrintUserCmdText(client, L"OK");
+	}
+
+	void BaseRmAllyFac(uint client, const wstring &args)
+	{
+		PlayerBase *base = GetPlayerBaseForClient(client);
+		if (CheckForBase(base, client)) return;
+
+		uint tag = 0;
+		try
+		{
+			tag = std::stoi(GetParam(args, ' ', 2));
+		}
+		catch (exception)
+		{
+			PrintUserCmdText(client, L"ERR No tag");
+			return;
+		}
+
+		if (base->ally_factions.find(tag) == base->ally_factions.end())
+		{
+			PrintUserCmdText(client, L"ERR Tag does not exist");
+			return;
+		}
+
+		base->ally_factions.erase(tag);
+		base->Save();
+		PrintUserCmdText(client, L"OK");
+	}
+
+	class Affiliations
+	{
+		class AffCell
+		{
+		public:
+			wstring nickname;
+			wstring factionname;
+			uint id;
+			AffCell(wstring a, wstring b, uint c)
+			{
+				nickname = a;
+				factionname = b;
+				id = c;
+			}
+			uint GetID() { return id; }
+		};
+
+		list<AffCell> AffList;
+
+		static bool IDComparision(AffCell & obj, int y)
+		{
+			if (obj.GetID() == y)
+				return true;
+			else
+				return false;
+		}
+
+		map<string, uint> factions;
+		void LoadListOfReps()
+		{
+			INI_Reader ini;
+
+			string factionpropfile = "..\\data\\initialworld.ini";
+			if (ini.open(factionpropfile.c_str(), false))
+			{
+				while (ini.read_header())
+				{
+					if (ini.is_header("Group"))
+					{
+						uint ids_name;
+						string nickname;
+						while (ini.read_value())
+						{
+							if (ini.is_value("nickname"))
+							{
+								nickname = ini.get_value_string();
+							}
+							else if (ini.is_value("ids_name"))
+							{
+								ids_name = ini.get_value_int(0);
+							}
+
+						}
+						factions[nickname] = ids_name;
+					}
+				}
+				ini.close();
+				ConPrint(L"Rep: Loaded %u factions\n", factions.size());
+			}
+		}
+
+		wstring GetFactioName(int ID)
+		{
+			try
+			{
+				wstring theaffiliation = HkGetWStringFromIDS(Reputation::get_name(ID)).c_str();
+				if (theaffiliation == L"Object Unknown")
+				{
+					theaffiliation = L"No Affiliation";
+				}
+				return theaffiliation;
+			}
+			catch (exception e)
+			{
+				return L"No Affiliation";
+			}
+		}
+
+		void LoadAffList()
+		{
+			if (AffList.size() == 0)
+			{
+				if (factions.size() == 0)
+					LoadListOfReps();
+				
+				for (map<string, uint>::iterator iter = factions.begin(); iter != factions.end(); iter++)
+				{
+					string factionnickname = iter->first;
+					uint ID = MakeId(factionnickname.c_str());

```

Comment:
Can we document what this does? Nothing else appears to use it. I assume this is like CreateID but for faction hashes?

Reply:
createIDTable = None
def FLFactionHash(nickName):
        FLFACTIONHASH_POLYNOMIAL = 0x1021

        # Build the crc lookup table if it hasn't been created
        global createIDTable
        if createIDTable is None:
                createIDTable = []
                for i in range(0, 2**8):
                        x = i << 8
                        for bit in range(0, 8):
                                x <<= 1
                                if x & 0x10000:
                                        x = (x ^ FLFACTIONHASH_POLYNOMIAL) & 0xFFFF
                        createIDTable.append(x)

        # Calculate the hash.
        hash = 0xFFFF
        for c in bytearray(nickName.lower().encode('utf-8')):
                hash = (hash >> 8) ^ createIDTable[(hash & 0xFF) ^ c]
        return hash

if __name__ == "__main__":
        import sys
        for n in sys.argv[1:]:
                h = FLFactionHash(n)
                
print(n, h, '0x{:04x}'.format(h))
alex@alex-laptop:~/game-repository (Development)$ python3 SERVICE/alexscripts/createfactionid.py fc_freelancer
fc_freelancer 4169 0x1049

Do you remember you gave me this function? Later I rewrote it to C++..... and a long time later after that Dartes told me about an already existing function in Flhook which does the same ;) He got the info from Laz. Basically MakeID is this function above, but in built in FLhook, same results.

Reply:
Provided comment to code
//MakeID function (in built in Flhook) is the same as mentioned here in C# to CreateFactionID https://github.com/DiscoveryGC/FLHook/blob/master/Plugins/Public/playercntl_plugin/setup_src/FLUtility.cs